include/micm/solver/solver.hpp

```diff
@@ -0,0 +1,38 @@
+/* Copyright (C) 2022 National Center for Atmospheric Research,
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#pragma once
+
+#include <cstddef>
+
+namespace micm
+{
+
+  /**
+   * @brief A base class to represent any time of solver
+   *
+   * @tparam T The underlying data format of the system
+   */
+  template<typename T>
+  class Solver
+  {
+   protected:
+    /// @brief The size of the system
+    std::size_t size_;
+
+   public:
+    /// @brief Default constructor
+    Solver();
+    /// @brief A virtual function to be defined by any solver baseclass
+    /// @param T The current species concentrations of the system
+    virtual void Solve(T) = 0;
```

Comment:
```suggestion
    virtual void Solve(std::vector<T>) = 0;
```

Reply:
If we're careful, it may not matter if it's a `double[]` or a `std::vector<T>`. I'm fine with enforcing one or the other, but I see no reason why we can't support both. Perhaps this is a policy decision?

Reply:
ah, ok - yeah, I like that better